cartographer/io/pbstream_info_main.cc

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <fstream>
+
+#include "cartographer/io/proto_stream.h"
+#include "cartographer/io/proto_stream_deserializer.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+
+DEFINE_string(pbstream_filename, "", "Proto stream file.");
+DEFINE_bool(all_debug_strings, false,
+            "Print debug strings of all serialized data.");
+
+namespace cartographer {
+namespace mapping {
+namespace {
+
+void Run(const std::string& pbstream_filename, bool all_debug_strings) {
+  LOG(INFO) << "Reading pbstream file from '" << pbstream_filename << "'...";
+  io::ProtoStreamReader reader(pbstream_filename);
+  io::ProtoStreamDeserializer deserializer(&reader);
+  const auto header = deserializer.header();
+  LOG(INFO) << "Header: " << header.DebugString();
+  for (const proto::TrajectoryBuilderOptionsWithSensorIds& trajectory_options :
+       deserializer.all_trajectory_builder_options()
+           .options_with_sensor_ids()) {
+    LOG(INFO) << "Trajectory options: " << trajectory_options.DebugString();
+  }
+  const proto::PoseGraph pose_graph = deserializer.pose_graph();
+  for (const proto::Trajectory& trajectory : pose_graph.trajectory()) {
+    LOG(INFO) << "Trajectory id: " << trajectory.trajectory_id()
+              << " has #nodes " << trajectory.node_size() << " has #submaps "
+              << trajectory.submap_size();
+  }
+  if (all_debug_strings) {
+    LOG(INFO) << "Pose graph: " << pose_graph.DebugString();
+  }
+
+  std::map<std::string, int> data_counts;
```

Comment:
unordered_map. Surprisingly enough, it can be faster than std::map for iterating over all elements  in some cases.

[benchmark](https://i.stack.imgur.com/27PqW.png)

Reply:
Done